src/main/java/org/hyperledger/fabric/sdk/security/CryptoPrimitives.java

```diff
@@ -1007,4 +984,173 @@ public byte[] certificateToDER(String certificatePEM) {
         return content;
     }
 
+
+    private static final long X509_RECHECK_MILLIS = TimeUnit.MINUTES.toMillis(30);
+
+    /**
+     * Cache for key for X.509 certificates.
+     */
+    private static class CertKey {
+        /** Pre-generated hash code for performance. */
+        final int hashCode;
+
+        /** The raw PEM data. */
+        final byte[] pemData;
+
+
+        /**
+         * New instance.
+         *
+         * @param pemData the PEM data
+         */
+        CertKey(byte[] pemData) {
+            // defensive copy
+            this.pemData = pemData.clone();
+
+            // pre-calculate the hash code
+            int h = 1;
+            for (byte b : pemData) {
+                h = (h * 31) + b;
+            }
+            hashCode = h;
```

Comment:
I literally copied that code from the Arrays class because I thought it wasn't in Java 1.8. I remember doing it. I have no idea why I thought a feature from Java 1.5 was not in Java 1.8. Too much time spent working on really old Java versions I guess. Thanks for spotting this.